swaybar/input.c

```diff
@@ -267,6 +277,116 @@ struct wl_pointer_listener pointer_listener = {
 	.axis_discrete = wl_pointer_axis_discrete,
 };
 
+static struct touch_slot *get_touch_slot(struct swaybar_touch *touch, int32_t id) {
+	int next = -1;
+	for (size_t i = 0; i < sizeof(touch->slots) / sizeof(*touch->slots); ++i) {
+		if (touch->slots[i].id == id) {
+			return &touch->slots[i];
+		}
+		if (next == -1 && !touch->slots[i].output) {
+			next = i;
+		}
+	}
+	if (next == -1) {
+		sway_log(SWAY_ERROR, "Ran out of touch slots");
+		return NULL;
+	}
+	return &touch->slots[next];
+}
+
+static void wl_touch_down(void *data, struct wl_touch *wl_touch,
+		uint32_t serial, uint32_t time, struct wl_surface *surface,
+		int32_t id, wl_fixed_t _x, wl_fixed_t _y) {
+	struct swaybar *bar = data;
+	struct swaybar_output *_output, *output;
+	wl_list_for_each(_output, &bar->outputs, link) {
+		if (_output->surface == surface) {
+			output = _output;
+			break;
+		}
+	}
+	if (!output) {
+		sway_log(SWAY_DEBUG, "Got touch event for unknown surface");
+		return;
+	}
+	struct touch_slot *slot = get_touch_slot(&bar->touch, id);
+	if (!slot) {
+		return;
+	}
+	slot->id = id;
+	slot->output = output;
+	slot->x = slot->start_x = wl_fixed_to_double(_x);
+	slot->y = slot->start_y = wl_fixed_to_double(_y);
+	slot->time = time;
+}
+
+static void wl_touch_up(void *data, struct wl_touch *wl_touch,
+		uint32_t serial, uint32_t time, int32_t id) {
+	struct swaybar *bar = data;
+	struct touch_slot *slot = get_touch_slot(&bar->touch, id);
+	if (!slot) {
+		return;
+	}
+	if (time - slot->time < 500) {
+		// Tap, treat it like a pointer click
+		process_hotspots(slot->output, slot->x, slot->y, BTN_LEFT);
+	}
+	slot->output = NULL;
+}
+
+static void wl_touch_motion(void *data, struct wl_touch *wl_touch,
+		uint32_t time, int32_t id, wl_fixed_t x, wl_fixed_t y) {
+	struct swaybar *bar = data;
+	struct touch_slot *slot = get_touch_slot(&bar->touch, id);
+	if (!slot) {
+		return;
+	}
+	int prev_progress = (int)((slot->x - slot->start_x)
+			/ slot->output->width * 100);
+	slot->x = wl_fixed_to_double(x);
+	slot->y = wl_fixed_to_double(y);
+	// "progress" is a measure from 0..100 representing the fraction of the
+	// output the touch gesture has travelled, positive when moving to the right
+	// and negative when moving to the left.
+	int progress = (int)((slot->x - slot->start_x)
+			/ slot->output->width * 100);
+	if (abs(progress) % 10 < abs(prev_progress) % 10) {
```

Comment:
I'm not sure I get how this condition works. Why use a modulo?

Reply:
This is true each time your finger slides past a threshold of 10% of the width of the screen.

Reply:
Actually this is pretty weird, not sure what I was thinking here. Will revisit this

Reply:
Would following condition work?
```c
if (abs(progress - prev_progress) > 0)
```

Reply:
I think so, yep

Reply:
Actually, not sure it would. That would be checking that you are moving *fast* enough, not *far* enough.

Reply:
```c
if (abs(progress) > 10)
```

?

Reply:
No, after further testing and reasoning I believe the original conditional is correct.

Reply:
Well I still don't understand it.

Reply:
![](https://sr.ht/8tV9.jpg)

Progress is a value from `[0..100)` which represents how far your finger has travelled from its starting point to the starting point plus the width of the screen. We subdivide the bar into 10 segments, and passing from one segment to the next causes the workspace to change. We offset these by your starting position, so if you touch, then travel 10% of the screen, you go to workspace +1, then 10% more and workspace +2, and so on.

`progress` will be `< prev_progress` when you cross one of these thresholds - the previous value will be near the end of the last segment, so a higher number (in this example 9), and the new value will be near the start of the next segment, so a lower number (in this example 2).

Reply:
Actually, reworked this a bit to make it behave a bit more predictably. Does the new approach make more sense?